File: src/main/java/com/justinnguyenme/base64image/Base64Image.java
package com.justinnguyenme.base64image;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * Created by Nguyen Hoang Anh Nguyen on 2/16/16.
 */
public class Base64Image {

    static final Handler HANDLER = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                default:
                    throw new AssertionError("Unknown handler message received: " + msg.what);
            }
        }
    };

    private static volatile Base64Image singleton = null;
    final ExecutorService service;
    private final Context context;

    private Base64Image(Context context, ExecutorService service) {
        this.context = context;
        this.service = service;
    }

    public static Base64Image with(Context context) {
        if (singleton == null) {
            synchronized (Base64Image.class) {
                if (singleton == null) {

                    float scaleFactor = 0.5f;
                    int cpuCount = Runtime.getRuntime().availableProcessors();
                    int maxThreads = (int) (cpuCount * scaleFactor);
                    maxThreads = (maxThreads > 0 ? maxThreads : 1);

                    final ExecutorService service = new ThreadPoolExecutor(
                            maxThreads, // core thread pool size
                            maxThreads, // maximum thread pool size
                            60, TimeUnit.SECONDS,
                            new LinkedBlockingDeque<Runnable>(),
                            new ThreadPoolExecutor.CallerRunsPolicy());

                    singleton = new Base64Image(context, service);
                }
            }
        }
        return singleton;
    }

    public RequestDecode decode(String base64) {
        if (base64.trim().length() == 0) {
            throw new IllegalArgumentException("Base64 must not be empty.");
        }
        return new RequestDecode(this, base64);
    }

    public RequestEncode encode(Bitmap bitmap) {
        if (bitmap == null) {
            throw new IllegalArgumentException("Bitmap must not be empty.");
        }
        return new RequestEncode(this, bitmap);
    }
}

File: src/main/java/com/justinnguyenme/base64image/Base64Image.kt
package com.justinnguyenme.base64image

import android.graphics.Bitmap
import java.util.concurrent.LinkedBlockingDeque
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit


class Base64Image private constructor() {

    private object Holder {
        val INSTANCE = Base64Image()
    }

    companion object {
        private const val SCALE = 0.5F
        private const val KEEP_ALIVE_TIME = 60L

        val instance: Base64Image by lazy { Holder.INSTANCE }
    }

    private val service: ThreadPoolExecutor

    init {
        val cpuCount = Runtime.getRuntime().availableProcessors()
        val thread = (cpuCount * SCALE).toInt()
        val maxThreads = if (thread > 0) thread else 1

        service = ThreadPoolExecutor(
                maxThreads, // core thread pool size
                maxThreads, // maximum thread pool size
                KEEP_ALIVE_TIME,
                TimeUnit.SECONDS,
                LinkedBlockingDeque(),
                ThreadPoolExecutor.CallerRunsPolicy())
    }

    fun encode(bitmap: Bitmap, function: (base64: String?) -> Unit) {
        val job = Encoder(bitmap, object : Listener<String?> {
            override fun onRetry(count: Int) {
            }

            override fun onStart() {
            }

            override fun onCompleted(result: String?) {
                function.invoke(result)
            }
        })

        service.submit(job)
    }

    fun decode(base64: String, function: (bitmap: Bitmap?) -> Unit) {
        val job = Decoder(base64, object : Listener<Bitmap?> {
            override fun onRetry(count: Int) {
            }

            override fun onStart() {
            }

            override fun onCompleted(result: Bitmap?) {
                function.invoke(result)
            }
        })

        service.submit(job)
    }
}